src/graph/LookupExecutor.cpp

```diff
@@ -0,0 +1,707 @@
+                prop = *aExpr->prop();
+                filters_.emplace_back(std::make_pair(prop, rExpr->op()));
+            } else {
+                return Status::SyntaxError("Unsupported expression ：%s",
+                                           rExpr->toString().c_str());
+            }
+            break;
+        }
+        case nebula::Expression::kFunctionCall : {
+            auto* fExpr = dynamic_cast<const FunctionCallExpression*>(expr);
+            auto* name = fExpr->name();
+            if (*name == "udf_is_in") {
+                return Status::SyntaxError("Unsupported function ： %s", name->c_str());
+            }
+            break;
+        }
+        default : {
+            return Status::SyntaxError("Syntax error ： %s", expr->toString().c_str());
+        }
+    }
+    return Status::OK();
+}
+
+Status LookupExecutor::checkFilter() {
+    auto status = traversalExpr(sentence_->whereClause()->filter());
+    if (!status.ok()) {
+        return status;
+    }
+    if (filters_.empty()) {
+        return Status::SyntaxError("Where clause error . have not index matching");
+    }
+    return Status::OK();
+}
+
+Status
+LookupExecutor::findValidIndex() {
+    std::vector<std::shared_ptr<nebula::cpp2::IndexItem>> indexes;
+    std::set<std::string> filterCols;
+    for (auto& filter : filters_) {
+        filterCols.insert(filter.first);
+    }
+    /**
+     * step 1 : found out all valid indexes. for example :
+     * tag (col1 , col2, col3)
+     * index1 on tag (col1, col2)
+     * index2 on tag (col2, col1)
+     * index3 on tag (col3)
+     *
+     * where where clause is below :
+     * col1 > 1 and col2 > 1 --> index1 and index2 are valid.
+     * col1 > 1 --> index1 is valid.
+     * col2 > 1 --> index2 is valid.
+     * col3 > 1 --> index3 is valid.
+     */
+    for (auto& index : indexes_) {
+        bool matching = true;
+        size_t filterNum = 1;
+        for (const auto& field : index->get_fields()) {
+            auto it = std::find_if(filterCols.begin(), filterCols.end(),
+                                   [field](const auto &name) {
+                                       return field.get_name() == name;
+                                   });
+            if (it == filterCols.end()) {
+                matching = false;
+                break;
+            }
+            if (filterNum++ == filterCols.size()) {
+                break;
+            }
+        }
+        if (!matching || index->get_fields().size() < filterCols.size()) {
+            continue;
+        }
+        indexes.emplace_back(index);
+    }
+
+    if (indexes.empty()) {
+        return Status::IndexNotFound();
+    }
+
+    /**
+     * step 2 , if have multiple valid indexes, get the best one.
+     * for example : if where clause is :
+     * col1 > 1 and col2 > 1 --> index1 and index2 are valid. get one of these at random.
+     * col1 > 1 and col2 == 1 --> only index2 is valid. because col2 have a equivalent value.
```

Comment:
I think both index1 and index2 are valid too in this case.

Reply:
> I think both index1 and index2 are valid too in this case.

Sure, the both indexes are valid, but I need choose a best one, because only need one index in storage layer.

Reply:
Update the comments.